src/components/container.js

```diff
@@ -279,27 +278,46 @@ function createFunctionContainer({ displayName, override } = {}) {
       ? store === override.Store
       : store.containedBy === FunctionContainer;
 
-  function FunctionContainer({ children, scope, isGlobal, ...restProps }) {
+  function FunctionContainer(props) {
+    const { children, ...restProps } = props;
+    const { scope, isGlobal, ...subProps } = restProps;
     const ctx = useContext(Context);
     const registry = useRegistry(scope, isGlobal, ctx);
+
+    // Store props in a ref to avoid re-binding actions when they change and re-rendering all
+    // consumers unnecessarily. The update is handled by an effect on the component instead
+    const propsRef = useRef({ prev: null, next: restProps, sub: subProps });
+    propsRef.current = {
+      prev: propsRef.current.next,
+      next: restProps,
+      sub: subProps, // TODO remove on next major
+    };
+
     const [containedStores, getContainedStore] = useContainedStore(
       scope,
       registry,
-      restProps,
+      propsRef,
       check,
       override
     );
+
+    // Use a stable object as is passed as value to context Provider
     const api = useApi(check, getContainedStore, ctx);
 
     // This listens for custom props change, and so we trigger container update actions
-    // before the re-render gets to consumers, hence why memo and not effect
-    useMemo(() => {
-      containedStores.forEach(({ handlers }) => {
-        handlers.onContainerUpdate?.();
-      });
-      // Deps are dynamic because we want to notify on any custom prop change
-      // eslint-disable-next-line react-hooks/exhaustive-deps
-    }, Object.values(restProps).concat(containedStores));
+    // before the re-render gets to consumers (hence why memo call on render).
+    // Also we use our own memoize instead of relying on memo itself because number of restProps
+    // might change and react throws if deps array length changes :/
+    useMemo(
+      () =>
+        memoize((stores, nextProps) => {
+          // no need to check if first render as stores will be empty anyway
+          stores.forEach(({ handlers }) => {
+            handlers.onContainerUpdate?.(nextProps, propsRef.current.prev);
+          });
+        }, true),
+      []
```

Comment:
The wrapping `useMemo` means that our memoize helper is slower than creating a new array + arrow function + React comparison. If you feel that's the case, fine but might be unnecessary. When I check our version, it also allocates an array, creates a function when called. So I think `useMemo` is just introducing an additional comparison step.

Let me know if my thinking is flawed, but removing memo would simplify this a lot :)

Reply:
> our memoize helper is slower than creating a new array + arrow function + React comparison

Not sure I follow. Where would you store the result of the memoization? Are you thinking of something like this to avoid extra array allocation and check?

```js
const [usePropsChangeEffect] = useState(() => memoize(...))
usePropsChangeEffect(containedStores, restProps);
```

Reply:
What I am suggesting is just `memoize(...)(containedStores, restProps)`. Should not matter if it's `useMemo`ed or not if not it is marginally slower.

Reply:
Ok, I think I got what you meant. Replaced the memo with a plain if and shallowEqual check